docs/metrics-data-model.md

```diff
@@ -0,0 +1,108 @@
+# k6 Time Series
+
+We want to introduce the time series concept to k6 for getting the benefit of efficiently identifying the combination between a metric and the relative tags.
+
+A summary of the benefits it brings across the platform:
+
+* It allows executing faster lookups and comparisons by a single value, the ID (hash) of the series.
+* It reduces the storage by allocating just one time a possible combination (the time series) and referencing it in all other places. 
+* It allows aggregating data by time series.
+* It enables easier detection of high cardinality issues just by counting the entire unique data set of generated time series.
+* It enables time series based outputs to simplify the integration's logic (e.g. Prometheus).
+
+## Data model
+
+#### Metrics and Series
+
+* Metric (name, type)
+* Time Series (metric, tags)
+* Sample (timestamp, value, time series ref)
+* Tag - (aka [k6/1831](https://github.com/grafana/k6/issues/1831))
+
+```go
+// This is the current k6 Metric polished from the Thresholds dependencies.
+type Metric struct {
+	name     string
+	type     MetricType
+	contains ValueType
+}
+
+type Tag struct {
+    Key, Value string
+}
+
+type TimeSeries struct {
+	id uint64 // hash(metric+tags)
+	metricName string
+	tags   []Tag
+	keys map[string]struct{}
+}
+
+type Sample struct {
+        TimeSeries *TimeSeries
+        Timestamp uint64
+        Value float64
+}
+```
+
+> **Note**: Tag and Metric without Sink+Threholds are represented here just as a general overview of the potential final data model. Tag is to be considered as [k6/1831](https://github.com/grafana/k6/issues/1831) and it isn't expected to be addressed in the same iteration.
+
+#### IDs
+
+The series IDs are `uint64` values generated by hashing the metric name and the tags' key-value pairs.
+
+#### Sink
+
+The sinks are implemented by metric types and they keep the series values up to date:
+
+* Counter:  a monotonic increasing value
+* Gauge: the latest value
+* SparseHistogram: counters per dynamic ranges (buckets) - https://grafana.com/blog/2021/11/03/how-sparse-histograms-can-improve-efficiency-precision-and-mergeability-in-prometheus-tsdb
+
+## Storage
+
+#### Time Series database (aka tsdb)
+
+We need to store all the time series generated from `k6` run during the execution so the other components (mostly the outputs) could query the storage for getting the value of the time series. It's expected to be in-memory and it should be concurrent-safe.
+
+```go
+type Repository interface {
```

Comment:
This will eventually live in the top [`metrics` k6 package](https://github.com/grafana/k6/tree/master/metrics), right? And it doesn't really store all that much data now that TimeSeries doesn't have built-in sink?

If so, then `Repository` may be a bit misleading, it's more like a `Registry` for time series, it is very similar to the `metrics.Registry` we already have for metrics: https://github.com/grafana/k6/blob/master/metrics/registry.go

So, maybe `TimeSeriesRegistry`? Or just `TSRegistry`? :thinking: Not sure I like these either :sweat_smile:

Reply:
hmm, or we can just add its few methods to the already existing `metrics.Registry` :thinking: They are very related and we already thread that `Registry` everywhere, and the final "interface" will look fairly tidy and logical, something like this:
```go
type Repository interface {
    NewMetric(name string, typ MetricType, t ...ValueType) (*Metric, error)
    MustNewMetric(name string, typ MetricType, t ...ValueType) *Metric
    Get(name string) *Metric

    GetOrCreateSeries(tags []Tag, metric *Metric) *TimeSeries // you can query the hash (ID) from the returned *TimeSeries
    // Some other potentially useful methods, though I am not sure if they are needed:
    GetSeriesWithHash(id uint64) (*TimeSeries, error)
    GetSeriesWithMetric(*Metric) []*TimeSeries 
}
```

:man_shrugging: